File: src/commands/list.ts
import path from 'node:path';
import pc from 'picocolors';
import { listWorktrees, repoRoot } from '../git.js';
import { log } from '../ui.js';

export async function listCmd(): Promise<void> {
  const repo = await repoRoot();
  const entries = await listWorktrees(repo);
  const managedRoot = path.join(repo, '.lcc', 'worktrees');

  const rows = entries
    .map((wt) => ({
      ...wt,
      managed: wt.path.startsWith(managedRoot + path.sep),
    }))
    .filter((wt) => !wt.isMain);

  if (rows.length === 0) {
    log.dim('No worktrees.');
    return;
  }

  const widthBranch = Math.max(...rows.map((r) => (r.branch ?? r.head.slice(0, 8)).length), 6);
  for (const wt of rows) {
    const label = wt.branch ?? `${wt.head.slice(0, 8)} (detached)`;
    const flags: string[] = [];
    if (wt.locked) flags.push(pc.yellow('locked'));
    if (wt.prunable) flags.push(pc.red('prunable'));
    if (wt.managed) flags.push(pc.dim('lcc'));
    const tail = flags.length ? '  ' + flags.join(' ') : '';
    console.log(`${pc.cyan(label.padEnd(widthBranch))}  ${pc.dim(wt.path)}${tail}`);
  }
}

File: zig/src/commands/list.zig
//! `lcc list` — worktrees in the current repo.

const std = @import("std");
const app_mod = @import("../app.zig");
const ui = @import("../ui.zig");

pub fn run(app: app_mod.App) !void {
    const repo = try app.repo();
    const choices = try app_mod.worktreeChoices(app, repo);

    if (choices.len == 0) {
        app.ui.hint("No worktrees.", .{});
        return;
    }

    var width: usize = 6;
    for (choices) |choice| {
        const label = if (choice.entry.branch) |b| b else app_mod.shortHead(choice.entry.head);
        width = @max(width, ui.displayWidth(label));
    }

    for (choices) |choice| {
        const label = if (choice.entry.branch) |b|
            try app.gpa.dupe(u8, b)
        else
            try std.fmt.allocPrint(app.gpa, "{s} (detached)", .{app_mod.shortHead(choice.entry.head)});
        const padded = try std.fmt.allocPrint(app.gpa, "{f}", .{ui.pad(label, width)});

        app.ui.info("{f}  {f}{f}{f}{f}", .{
            ui.cyan(padded),
            ui.dim(choice.entry.path),
            ui.yellow(if (choice.entry.locked) "  locked" else ""),
            ui.red(if (choice.entry.prunable) "  prunable" else ""),
            ui.dim(if (choice.managed) "  lcc" else ""),
        });
    }
}
